### PR TITLE
Assignment view with approval features.

### PR DESCRIPTION
**Pull request by JackUrb**
Introduces a new view to the frontend that allows viewing assignments that have been worked on and are in any completed state. This works by leveraging the directory structure established in #1253 and the database from earlier. 

This was the table of info from before:
screen shot 2018-10-31 at 5 25 05 pm

### This is the new assignment view you can get to by clicking through an assignment id:
screen shot 2018-10-31 at 5 25 47 pm

### It scrapes task description instructions from the `task_config.py` file in the run's task folder (checking both `parlai` and `parlai_internal` folders).
screen shot 2018-10-31 at 5 26 03 pm

### It can find onboarding chats if they were saved for an assignment and knows to avoid them otherwise.
**Exist:**
screen shot 2018-10-31 at 5 53 10 pm
**Not exist:**
screen shot 2018-10-31 at 5 26 16 pm

### It has an approval flow at the bottom that allows approving and rejecting (with reason):
screen shot 2018-10-31 at 5 26 26 pm
screen shot 2018-10-31 at 5 26 46 pm

### The approval flow knows the current assignment state, (approved, rejected) and can override rejected HITs to acceptances.
screen shot 2018-10-31 at 5 27 24 pm
screen shot 2018-10-31 at 5 26 56 pm
screen shot 2018-10-31 at 5 27 14 pm

### It also has a bonus modal:
screen shot 2018-10-31 at 5 28 04 pm

### And a block modal:
screen shot 2018-10-31 at 5 29 06 pm

### Unordered list of next steps
- Functionality for converting approved assignments into a ParlAI-Formatted dataset
- Better displays for the content in the tables
- Implement run status tracking
- Keep track of worker blocks to swap block to unblock for a worker
- Context displays (requires moving frontend to React)
- Provide the block option in the workers page for a worker
- Create a view that allows stepping through 
- Add qualification tracking
- Add a view for testing hits without needing to launch heroku

### CODE REVIEW

**Comment by alexholdenmiller** [comment]
When you get the chance, please fix the unit test errors :)

File: parlai/mturk/core/mturk_data_handler.py [hunk 1]
```diff
@@ -405,17 +424,36 @@ def log_start_task(self, worker_id, assignment_id, conversation_id):
...
     def log_award_amount(self, worker_id, assignment_id, amount, reason):
-        """Update a pairing state to add a bonus to be paid, appends reason"""
+        """Update a pairing state to add a task bonus to be paid,
+        appends reason. To be used for automatic evaluation bonuses
+        """
         with self.table_access_condition:
             conn = self._get_connection()
             c = conn.cursor()
             reason = "${} for {}\n".format(amount, reason)
+            amount = int(amount * 100)
```
Comment (alexholdenmiller) [inline parlai/mturk/core/mturk_data_handler.py]: maybe include an inline comment here explaining why you're scaling the amount 100x  😂 to a casual observer it looks magic

**Comment by JackUrb** [comment]
@stephenroller is helping me out with that one on the travis side 👍

File: parlai/mturk/webapp/server.py [hunk 2]
```diff
@@ -304,16 +319,158 @@ def get(self, worker_target):
...
+        Requires a reason, assignment_id, a unique token (for idempotence),
+        and the bonus amount IN CENTS
+        """
+        data = tornado.escape.json_decode(self.request.body)
+        reason = data['reason']
+        assignment_id = data['assignment_id']
+        bonus_cents = data['bonus_cents']
+        token = data['bonus_token']
+
+        amount = bonus_cents / 100.0
```
Comment (alexholdenmiller) [inline parlai/mturk/webapp/server.py]: same here I think even just  `# cents => dollars` or change `amount` var to `dollars` or `dollar_amount`

File: parlai/mturk/core/mturk_data_handler.py (hunk 1 above)
Comment (JackUrb) [reply parlai/mturk/core/mturk_data_handler.py]: Ah of course - don't want to mislead people into thinking we're just going to multiply bonuses by 100x 😂

File: parlai/mturk/core/mturk_data_handler.py [hunk 3]
```diff
@@ -405,15 +426,36 @@ def log_start_task(self, worker_id, assignment_id, conversation_id):
...
-        """Update a pairing state to add a bonus to be paid, appends reason"""
+        """Update a pairing state to add a task bonus to be paid,
+        appends reason. To be used for automatic evaluation bonuses
+        """
         with self.table_access_condition:
             conn = self._get_connection()
             c = conn.cursor()
             reason = "${} for {}\n".format(amount, reason)
+            # Bonus amount is stored in a cents int in the SQL table
+            cent_amount = int(amount * 100)
```
Comment (klshuster) [inline parlai/mturk/core/mturk_data_handler.py]: do we enforce somewhere else that bonuses are set in dollar amounts?

edit:  i notice that is the case for the webapp, curious if this is still enforced elsewhere (or if that even matters for this particular function)

File: parlai/mturk/core/mturk_data_handler.py [hunk 4]
```diff
@@ -669,3 +727,39 @@ def get_all_task_pairings_for_worker(self, worker_id, task_group_id=None):
                       (worker_id, task_group_id))
             results = c.fetchall()
             return results
+
+    @staticmethod
+    def get_conversation_data(task_group_id, conv_id, worker_id, is_sandbox):
```
Comment (klshuster) [inline parlai/mturk/core/mturk_data_handler.py]: for some reason, this way of returning the data seems very satisfying to me

File: parlai/mturk/webapp/dev/app.jsx [hunk 5]
```diff
@@ -34,6 +35,93 @@ function resolveState(state_string) {
...
+function postData(url = ``, data = {}) {
+  // Default options are marked with *
+  return fetch(url, {
+      method: "POST", // *GET, POST, PUT, DELETE, etc.
+      mode: "cors", // no-cors, cors, *same-origin
+      cache: "no-cache", // *default, no-cache, reload, force-cache, only-if-cached
+      credentials: "same-origin", // include, *same-origin, omit
+      headers: {
+          "Content-Type": "application/json; charset=utf-8",
+          // "Content-Type": "application/x-www-form-urlencoded",
```
Comment (klshuster) [inline parlai/mturk/webapp/dev/app.jsx]: nit: delete comment?

File: parlai/mturk/webapp/dev/app.jsx [hunk 6]
```diff
@@ -543,6 +638,679 @@ class RunPanel extends React.Component {
   }
 }
 
+class AssignmentInstructions extends React.Component {
+  render() {
+    let instructions = this.props.data;
```
Comment (klshuster) [inline parlai/mturk/webapp/dev/app.jsx]: is this similar to swift, where `let` gives you an immutable/constant ref?

File: parlai/mturk/webapp/server.py [hunk 7]
```diff
@@ -304,16 +319,158 @@ def get(self, worker_target):
...
+            'task': MTurkDataHandler.get_conversation_data(
+                run_id, conversation_id, worker_id, self.state['is_sandbox']),
+        }
+
+        # Get assignment instruction html
+        taskname = '_'.join(run_id.split('_')[:-1])
+        find_location = 'parlai.mturk.tasks.{}.task_config'.format(taskname)
+        find_location_internal = \
+            'parlai_internal.mturk.tasks.{}.task_config'.format(taskname)
+        try:
```
Comment (klshuster) [inline parlai/mturk/webapp/server.py]: suppose you have two versions of a task, one internal and one external (e.g. you're extending an external task internally for a project), is there a way to specify overriding this order and using the internal task instead?

File: parlai/mturk/core/mturk_data_handler.py (hunk 3 above)
Comment (JackUrb) [reply parlai/mturk/core/mturk_data_handler.py]: We don't necessarily enforce it, though this call is only made from MTurkManager, and that API suggests the bonus amount is in dollars. The cent representation is only ever seen on the backend and it only exists for SQL storage reasons.

File: parlai/mturk/core/mturk_data_handler.py (hunk 4 above)
Comment (JackUrb) [reply parlai/mturk/core/mturk_data_handler.py]: Once the data is in an organized place, it's really easy to access it. The accessors along the way make sure that if someone manually alters the folder structure there is no catastrophic failure and I can give a good reason.

File: parlai/mturk/webapp/dev/app.jsx (hunk 5 above)
Comment (JackUrb) [reply parlai/mturk/webapp/dev/app.jsx]: Ooops this was copypasta from the fetch api docs

File: parlai/mturk/webapp/dev/app.jsx (hunk 6 above)
Comment (JackUrb) [reply parlai/mturk/webapp/dev/app.jsx]: let is limited in scope to the current block. I should be moving more of my `var`s to be `let`s, but I haven't taken the time to clean yet.

File: parlai/mturk/webapp/server.py (hunk 7 above)
Comment (JackUrb) [reply parlai/mturk/webapp/server.py]: Right now I haven't added any flag to do such a thing. I'm unsure how frequently we'd run into a problem like this though, as usually things that end up public aren't collecting any more data, and if you're making an extension on a task it'll have an extended name (usually).

**Comment by JackUrb** [comment]
Merging with build failures because travis is having issues binding sockets today.

**Pull request merged**